extension/wallet/src/protocol/extract_trade.rs

```diff
@@ -1,38 +1,54 @@
-use crate::{
-    wallet::{compute_balances, current, get_txouts, Wallet},
-    TradeSide,
-};
-use anyhow::{bail, Context, Result};
+use crate::esplora::EsploraClient;
+use crate::wallet::current;
+use crate::{TradeSide, ESPLORA_API_URL, LOADED_WALLET};
 use elements::{confidential, secp256k1_zkp::SECP256K1, Transaction, TxOut};
-use futures::lock::Mutex;
 use itertools::Itertools;
 use serde::{Deserialize, Serialize};
+use wasm_bindgen::prelude::*;
+
+/// Decomposes a transaction into:
+///
+/// - Sell amount, sell balance before and sell balance after.
+/// - Buy amount, buy balance before and buy balance after.
+///
+/// To do so we unblind confidential `TxOut`s whenever necessary.
+#[wasm_bindgen]
+pub async fn extract_trade(wallet_name: String, transaction: JsValue) -> Result<JsValue, JsValue> {
+    let transaction: Transaction = map_err_from_anyhow!(transaction.into_serde())?;
+    let trade = map_err_from_anyhow!(extract_trade_w(wallet_name, transaction).await)?;
+    let trade = map_err_from_anyhow!(JsValue::from_serde(&trade))?;
+
+    Ok(trade)
+}
 
 // TODO: Public APIs should return specific error struct/enum
-pub async fn extract_trade(
-    name: String,
-    current_wallet: &Mutex<Option<Wallet>>,
+pub async fn extract_trade_w(
+    wallet_name: String,
     transaction: Transaction,
-) -> Result<Trade> {
-    let wallet = current(&name, current_wallet).await?;
+) -> Result<Trade, Error> {
+    let client = {
+        let guard = ESPLORA_API_URL.lock().expect_throw("can get lock");
+        EsploraClient::new(guard.clone())
+    };
 
-    let txouts = get_txouts(&wallet, |utxo, txout| Ok(Some((utxo, txout)))).await?;
-    let balances = compute_balances(
-        &wallet,
-        &txouts
-            .iter()
-            .map(|(_, txout)| txout)
-            .cloned()
-            .collect::<Vec<_>>(),
-    );
+    let (utxos, balances, blinding_key, address) = {
+        let mut wallet = current(&wallet_name, &LOADED_WALLET)
+            .await
+            .map_err(Error::LoadWallet)?;
+        wallet.sync(&client).await.map_err(Error::SyncWallet)?;
 
-    let blinding_key = wallet.blinding_key();
+        let utxos = wallet.utxos();
+        let balances = wallet.compute_balances();
+        let blinding_key = wallet.blinding_secret_key();
+        let address = wallet.address();
+        (utxos, balances, blinding_key, address)
+    };
 
     let our_inputs = transaction
```

Comment:
Finding our inputs (and our outputs) in a transaction could be considered generic enough wallet functionality so we can include it in the wallet lib.

My goal would be to get rid of `Wallet::utxos`. I think it would be really good for encapsulation (and hence API evolvement) if we would not expose that.

Reply:
For example, if we ever want to change our wallet to not reuse addresses or change how blinding keys are related to addresses, this code would have to be updated in `waves`. But if the wallet is in control of the task of "identify my inputs and outputs", we can make the change transparently.